### PR TITLE
[Parallel] Fix system errors expected an instance of SystemError but got string

### LINKED ISSUE #6901: [Parallel] System errors may contain strings
# Bug Report

| Subject        | Details                                                         |
| :------------- | :---------------------------------------------------------------|
| Rector version | 0.12.10            |

When running rector with parallel, sometimes the following error is thrown.

```
In Assert.php line 1591:
  Expected an instance of this class or to this class among his parents Rector\Core\ValueObject\Error\SystemError. Got: string                            
```

The string is '**Reached system errors count limit of 20, exiting...**', code position:
https://github.com/rectorphp/rector/blob/8c823197becb3905e42b9d7f7849f1174e8b47cf/packages/Parallel/Application/ParallelFileProcessor.php#L159

## Minimal PHP Code Causing Issue

Sometime happen when running rector with parallel.

## Expected Behaviour

No assertion errors should occur.

### PR DESCRIPTION
**Pull request by zingimmick**
Fixes: https://github.com/rectorphp/rector/issues/6901

### CODE REVIEW

File: packages/ChangesReporting/Output/ConsoleOutputFormatter.php [hunk 1]
```diff
@@ -101,11 +101,15 @@ private function reportFileDiffs(array $fileDiffs): void
...
 
     /**
-     * @param SystemError[] $errors
+     * @param array<SystemError|string> $errors
      */
     private function reportErrors(array $errors): void
     {
         foreach ($errors as $error) {
+            if (! $error instanceof SystemError) {
+                $this->outputStyle->error($error);
```
Comment (samsonasik) [inline packages/ChangesReporting/Output/ConsoleOutputFormatter.php]: is it means that the error will still shown? what message error that will be shown?

**Comment by TomasVotruba** [comment]
The fix should happen in the other direction. Where the error message is string, should be object.

**Comment by zingimmick** [comment]
> The fix should happen in the other direction. Where the error message is string, should be object.

@TomasVotruba I tried to convert it to an object, but the required argument `$relativeFilePath` is missing.

The location of the code where the error message is string:

https://github.com/rectorphp/rector-src/blob/34c61ba4498150972db843fca807416540ca71e8/packages/Parallel/Application/ParallelFileProcessor.php#L169
https://github.com/rectorphp/rector-src/blob/34c61ba4498150972db843fca807416540ca71e8/packages/Parallel/Application/ParallelFileProcessor.php#L214
https://github.com/rectorphp/rector-src/blob/34c61ba4498150972db843fca807416540ca71e8/packages/Parallel/Application/ParallelFileProcessor.php#L224-L227

File: packages/ChangesReporting/Output/ConsoleOutputFormatter.php (hunk 1 above)
Comment (zingimmick) [reply packages/ChangesReporting/Output/ConsoleOutputFormatter.php]: This is for support the error message is string. If the error must be an object https://github.com/rectorphp/rector-src/pull/1610#issuecomment-1003574675, this change is not needed.
image

**Comment by TomasVotruba** [comment]
> but the required argument $relativeFilePath is missing.

It could be either optional or with with `AbstractSystemError`. How would that work?

The reason is we should work with strict objects, moreover when we juggle around json data in paralle processes. Only that way we can control what data types are on I/O. When we start to send strings, it could be anything, from error message, to file content, and we'll be unable to resolve what we work with. 

Jsonable value objects makes sure we know the types all the time.

**Comment by zingimmick** [comment]
> The reason is we should work with strict objects, moreover when we juggle around json data in paralle processes. Only that way we can control what data types are on I/O. When we start to send strings, it could be anything, from error message, to file content, and we'll be unable to resolve what we work with.

I see. I'll try the optional first and see what happens.

**Comment by zingimmick** [comment]
This is what an error message looks like with `relativeFilePath` optional.

```
 [ERROR] Could not process "" file, due to:                                                                             
         "Reached system errors count limit of 20, exiting...".    
```

**Comment by TomasVotruba** [comment]
How should we rephrase it?

File: packages/Parallel/Application/ParallelFileProcessor.php [hunk 2]
```diff
@@ -166,7 +166,7 @@ function (array $json) use (
                     // decode arrays to objects
                     foreach ($json[Bridge::SYSTEM_ERRORS] as $jsonError) {
                         if (is_string($jsonError)) {
-                            $systemErrors[] = 'System error: ' . $jsonError;
+                            $systemErrors[] = new SystemError('System error: ' . $jsonError);
```
Comment (staabm) [inline packages/Parallel/Application/ParallelFileProcessor.php:170]: Could you type this array, so phpstan would error when strings are pushed onto the array?

Comment (zingimmick) [reply packages/Parallel/Application/ParallelFileProcessor.php:170]: Add a type to this array, but phpstan does not error when a string is pushed into the array. 🤔

**Comment by zingimmick** [comment]
> How should we rephrase it?

How about
```
 [ERROR] Could not process some files, due to:                                                                             
         "Reached system errors count limit of 20, exiting...".    
```

File: packages/Parallel/Application/ParallelFileProcessor.php (hunk 2 above)
Comment (zingimmick) [reply packages/Parallel/Application/ParallelFileProcessor.php:170]: It's not supported yet. https://github.com/phpstan/phpstan/issues/3269, https://github.com/phpstan/phpstan/issues/3703

**Comment by TomasVotruba** [comment]
@zingimmick Sounds good :+1:

**Comment by TomasVotruba** [comment]
Thank you :+1: 

I'm merging this now, so we can fix important bug that blocks us at the moment. Feel free to send PR with better wording.

**Pull request merged**